scripts/dev/generate_config_types.py

```diff
@@ -0,0 +1,237 @@
+"""This script auto-generates the `qutebrowser/config/configcontainer.py` file."""
+
+from __future__ import annotations
+
+from pathlib import Path
+from textwrap import dedent
+from typing import Union, cast
+from collections.abc import Mapping
+from typing_extensions import TypeAlias, assert_never
+
+# pylint: disable=unused-import
+import qutebrowser.config.config  # noqa: F401, needed to avoid cyclical import errors
+# pylint: enable=unused-import
+
+from qutebrowser.config import configdata, configtypes
+from qutebrowser.config.configdata import Option
+
+
+NestedConfig: TypeAlias = dict[str, Union[Option, "NestedConfig"]]
+
+
+def make_nested_config(config: Mapping[str, Option]) -> NestedConfig:
+    """The original configdata.yml defines nested keys using `.`s in a flat tree.
+
+    This function returns a new dict where options are grouped and nested, e.g.
+    `'auto_save.session': Option(...)` -> `{'auto_save': {'session': Option(...)}}`
+    """
+    result = {}
+    for key, value in config.items():
+        parts = key.split(".")
+        current = result
+        for part in parts[:-1]:
+            current = current.setdefault(part, {})
+        current[parts[-1]] = value
+    return result
+
+
+def _get_type_hint(typ: configtypes.ConfigType) -> str:  # noqa: C901
```

Comment:
I don't think it's the right approach to have a big `if ...:` block in here (very far from where those config types are defined!), and then use `assert_never` here (also needing the big `ConfigType` union definition to work).

What do you think about adding a

```python
@abstractmethod
def py_type(self) -> str:
    raise NotImplementedError
```

to `configtypes.BaseType` instead? As per [the mypy docs](https://mypy.readthedocs.io/en/stable/class_basics.html#abstract-base-classes-and-multiple-inheritance), this should make mypy complain statically if it's not defined anywhere (and if not, we'd still notice as soon as the script is being run). I wouldn't go as far as doing `metaclass=ABCMeta`, I'm not a fan of the complexity and we don't need runtime checking at class instantiation.

That would mean:

- The definition of the Python types are very close to where the config types are defined, making it easier for people changing those, and reducing the possibility of missing it.
- It avoids having a giant `if ...:` in here, also see [Replace Conditional with Polymorphism](https://refactoring.guru/replace-conditional-with-polymorphism)
- We hopefully get static checking without all the overhead of defining `ConfigType` and using `assert_never`.

Reply:
agree in general this is a much better approach!

unfortunately it looks like mypy only complains about unimplemented abstract methods when the class is instantiated or when `@final` is used but I don't think we can really use `@final` here because classes like `String` are used as a parent class... 🤔

we don't get the protection from the class instantiation case because most of the classes are not constructed directly, so these are the only errors I get after adding the abstract method
```
❯ tox -e mypy-pyqt6
mypy-pyqt6: commands[0]> .tox/mypy-pyqt6/bin/python -m mypy --always-true=USE_PYQT6 --always-false=USE_PYQT5 --always-false=USE_PYSIDE6 --always-false=IS_QT5 --always-true=IS_QT6 --always-true=IS_PYQT --always-false=IS_PYSIDE qutebrowser
qutebrowser/config/configtypes.py: note: In member "__init__" of class "ListOrValue":
qutebrowser/config/configtypes.py:593: error: Cannot instantiate abstract class "List" with abstract attribute "py_type"  [abstract]
            self.listtype = List(valtype=valtype, none_ok=none_ok, **kwargs)
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
qutebrowser/config/configtypes.py: note: In member "__init__" of class "FlagList":
qutebrowser/config/configtypes.py:670: error: Cannot instantiate abstract class "String" with abstract attribute "py_type"  [abstract]
                valtype=String(),
                        ^~~~~~~~
qutebrowser/config/configtypes.py: note: In member "__init__" of class "ShellCommand":
qutebrowser/config/configtypes.py:1600: error: Cannot instantiate abstract class "String" with abstract attribute "py_type"  [abstract]
            super().__init__(valtype=String(), none_ok=none_ok, completions=completions)
                                     ^~~~~~~~
qutebrowser/config/configtypes.py: note: In member "__init__" of class "Padding":
qutebrowser/config/configtypes.py:1754: error: Cannot instantiate abstract class "String" with abstract attribute "py_type"  [abstract]
                keytype=String(),
                        ^~~~~~~~
qutebrowser/config/configtypes.py:1755: error: Cannot instantiate abstract class "Int" with abstract attribute "py_type"  [abstract]
                valtype=Int(minval=0, none_ok=none_ok),
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 5 errors in 1 file (checked 209 source files)
```
actually, how do you think this `.py_type()` implementation would work with the `.get_valid_values()` handling in the current implementation? would that need to be duplicated for every type that can have a list of valid values? hmm

Reply:
> we don't get the protection from the class instantiation case because most of the classes are not constructed directly, so these are the only errors I get after adding the abstract method

Ah, that's unfortunate. It looks like pylint is already handling this nicely, however, even without the `@abstractmethod` decorator:

<details>

```
************* Module qutebrowser.config.configtypes
qutebrowser/config/configtypes.py:343:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'MappingType' (abstract-method)
qutebrowser/config/configtypes.py:377:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'String' (abstract-method)
qutebrowser/config/configtypes.py:454:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'UniqueCharString' (abstract-method)
qutebrowser/config/configtypes.py:473:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'List' (abstract-method)
qutebrowser/config/configtypes.py:567:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'ListOrValue' (abstract-method)
qutebrowser/config/configtypes.py:646:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'FlagList' (abstract-method)
qutebrowser/config/configtypes.py:716:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Bool' (abstract-method)
qutebrowser/config/configtypes.py:756:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'BoolAsk' (abstract-method)
qutebrowser/config/configtypes.py:860:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Int' (abstract-method)
qutebrowser/config/configtypes.py:882:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Float' (abstract-method)
qutebrowser/config/configtypes.py:907:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Perc' (abstract-method)
qutebrowser/config/configtypes.py:940:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'PercOrInt' (abstract-method)
qutebrowser/config/configtypes.py:1025:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Command' (abstract-method)
qutebrowser/config/configtypes.py:1049:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'ColorSystem' (abstract-method)
qutebrowser/config/configtypes.py:1061:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'IgnoreCase' (abstract-method)
qutebrowser/config/configtypes.py:1075:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'QtColor' (abstract-method)
qutebrowser/config/configtypes.py:1146:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'QssColor' (abstract-method)
qutebrowser/config/configtypes.py:1225:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Font' (abstract-method)
qutebrowser/config/configtypes.py:1256:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'FontFamily' (abstract-method)
qutebrowser/config/configtypes.py:1280:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Regex' (abstract-method)
qutebrowser/config/configtypes.py:1355:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Dict' (abstract-method)
qutebrowser/config/configtypes.py:1466:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'File' (abstract-method)
qutebrowser/config/configtypes.py:1507:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Directory' (abstract-method)
qutebrowser/config/configtypes.py:1532:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'FormatString' (abstract-method)
qutebrowser/config/configtypes.py:1578:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'ShellCommand' (abstract-method)
qutebrowser/config/configtypes.py:1622:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Proxy' (abstract-method)
qutebrowser/config/configtypes.py:1680:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'SearchEngineUrl' (abstract-method)
qutebrowser/config/configtypes.py:1710:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'FuzzyUrl' (abstract-method)
qutebrowser/config/configtypes.py:1738:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Padding' (abstract-method)
qutebrowser/config/configtypes.py:1768:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Encoding' (abstract-method)
qutebrowser/config/configtypes.py:1785:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Position' (abstract-method)
qutebrowser/config/configtypes.py:1797:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'TextAlignment' (abstract-method)
qutebrowser/config/configtypes.py:1808:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'ElidePosition' (abstract-method)
qutebrowser/config/configtypes.py:1820:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'VerticalPosition' (abstract-method)
qutebrowser/config/configtypes.py:1833:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Url' (abstract-method)
qutebrowser/config/configtypes.py:1851:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'SessionName' (abstract-method)
qutebrowser/config/configtypes.py:1866:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'SelectOnRemove' (abstract-method)
qutebrowser/config/configtypes.py:1888:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'ConfirmQuit' (abstract-method)
qutebrowser/config/configtypes.py:1932:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'NewTabPosition' (abstract-method)
qutebrowser/config/configtypes.py:1949:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'LogLevel' (abstract-method)
qutebrowser/config/configtypes.py:1963:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'Key' (abstract-method)
qutebrowser/config/configtypes.py:1987:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'UrlPattern' (abstract-method)
qutebrowser/config/configtypes.py:2012:0: W0223: Method 'py_type' is abstract in class 'BaseType' but is not overridden in child class 'StatusbarWidget' (abstract-method)
```

</details>



> actually, how do you think this `.py_type()` implementation would work with the `.get_valid_values()` handling in the current implementation? would that need to be duplicated for every type that can have a list of valid values? hmm

Fair point. Maybe it should just default to that behavior (raising an exception if there are no `valid_values`), and then subclasses override it as needed? That would not give us static checking for exhaustiveness anymore, but I don't think that should be a priority here either. A test in `TestAll` from `tests/unit/config/test_configtypes.py` could just call all `py_type()` methods and ensure they don't raise.

Reply:
ahhhh I had no idea pylint would report errors like that, nice!